runtime/shape.js

```diff
@@ -123,6 +123,18 @@ ${this._slotsToManifestString()}
     return new Shape(this.name, handles, slots);
   }
 
+  canEnsureResolved() {
+    for (let typeVar of this._typeVars)
+      if (!typeVar.object[typeVar.field].canEnsureResolved()) return false;
+    return true;
+  }
+
+  maybeEnsureResolved() {
+    for (let typeVar of this._typeVars)
+      if (!typeVar.object[typeVar.field].maybeEnsureResolved()) return false;
```

Comment:
if we persist with maybeEnsureResolved() being something that we expect could fail, then this should be able to unwind on a failure partway through.

In practice I guess that means cloning each typeVar, maybeEnsureResolving the clones, failing when indicated, then either replacing or just maybeEnsureResolving the originals

Reply:
Done, I think. Can you PTAL?